include/gz/math/CoordinateVector3.hh

```diff
@@ -0,0 +1,276 @@
+    public: CoordinateVector3 operator+(const CoordinateVector3 &_v) const;
+
+    /// \brief Addition assignment operator
+    /// \param[in] _v vector to add
+    /// \return the sum vector
+    /// \throws std::bad_variant_access if the vectors are not both same type
+    public: const CoordinateVector3 &operator+=(const CoordinateVector3 &_v);
+
+    /// \brief Negation operator
+    /// \return negative of this vector
+    public: CoordinateVector3 operator-() const;
+
+    /// \brief Subtraction operators
+    /// \param[in] _pt a vector to subtract
+    /// \return a vector after the subtraction
+    /// \throws std::bad_variant_access if the vectors are not both same type
+    public: CoordinateVector3 operator-(
+        const CoordinateVector3 &_pt) const;
+
+    /// \brief Subtraction assignment operators
+    /// \param[in] _pt subtrahend
+    /// \return a vector after the subtraction
+    /// \throws std::bad_variant_access if the vectors are not both same type
+    public: const CoordinateVector3 &operator-=(const CoordinateVector3 &_pt);
+
+    /// \brief Equality test with tolerance.
+    /// \param[in] _v the vector to compare to
+    /// \param[in] _tol equality tolerance for metric components.
+    /// \param[in] _ang_tol equality tolerance for spherical components.
+    /// \return true if the vectors are equal within the tolerance specified by
+    ///         _tol and _ang_tol.
+    public: bool Equal(const CoordinateVector3 &_v,
+                       const double &_tol,
+                       const math::Angle &_ang_tol) const;
+
+    /// \brief Equal to operator
+    /// \param[in] _v The vector to compare against
+    /// \return true if the vectors are equal within a default tolerance (1e-3),
+    ///         false otherwise
+    public: bool operator==(const CoordinateVector3 &_v) const;
+
+    /// \brief Not equal to operator
+    /// \param[in] _v The vector to compare against
+    /// \return false if the vectors are not equal within a default tolerance
+    ///         (1e-3), true otherwise
+    public: bool operator!=(const CoordinateVector3 &_v) const;
+
+    /// \brief See if all vector components are finite (e.g., not nan)
+    /// \return true if is finite or false otherwise
+    public: bool IsFinite() const;
+
+    /// \brief Equality test
+    /// \remarks This is equivalent to the == operator
+    /// \param[in] _v the other vector
+    /// \return true if the 2 vectors have the same values, false otherwise
+    public: bool Equal(const CoordinateVector3 &_v) const;
+
+    /// \brief Get the x value of a metric vector.
+    /// \return The x component of the metric vector.
+    /// \throws std::bad_variant_access if the vector is spherical.
+    public: double X() const;
+
+    /// \brief Get the latitude of a spherical vector.
+    /// \return The latitude of the spherical vector.
+    /// \throws std::bad_variant_access if the vector is metric.
```

Comment:
Yes, this is a part of the API I didn't like, too. I was thinking about creating a custom exception to completely hide the std::variant in PIMPL.

I agree exceptions aren't good in general, but in this API, they'd be thrown only when the program is written wrongly (e.g. a spherical value is passed to a method that expects metric).

`std::optional` is definitely an option and I'll try to convert the exceptions to this approach.

An even better design would be using `std::expected<CoordinateVector3, std::string>` to allow not only saying that the operation failed, but also providing a user-readable reason. However, `std::expected` is not available in C++17. There is a header-only single-file shim provided by https://github.com/TartanLlama/expected which I use for example here https://github.com/ctu-vras/ros-utils/blob/1b66e6d64d237af6411225bcd62822946f436d9e/image_transport_codecs/include/image_transport_codecs/image_transport_codecs.h#L255 . I really like this approach as it provides enough debugging information, but doesn't come with the performance penalty of exceptions. The question is, could the shim library be sneaked into GZ libraries (probably in a [custom namespace to avoid naming collisions](https://github.com/ctu-vras/ros-utils/blob/1b66e6d64d237af6411225bcd62822946f436d9e/cras_cpp_common/include/cras_cpp_common/expected.hpp))? Which one? Utils? And is that doable in the available time frame?

Reply:
I'm actually not sure. I don't know what the ramifications of a CC-0 license is. @azeey can we include it in our code base? `std::expected` would be really nice and future proof also.

Reply:
Should be okay license-wise: https://www.gnu.org/licenses/license-compatibility.html

> We say that a license A subsumes license B when compliance with license A implies compliance with license B.

> The Apache 2.0 license subsumes the BSD, Expat, X11, ISC and CC-0 licenses.

Reply:
I don't think we have enough time to introduce another dependency.  I was recently working on enforcing an input range in `gz::math::Color` and ran into the same issue. I think the only option we have right now if it's in a constructor is to print an error message, but I'd like to figure out a more proper solution like using `std::expected` in the future. For other functions that return a value, `std::optional` is a good choice and used a lot in the Gazebo codebase.